jdaviz/utils.py

```diff
@@ -295,3 +297,46 @@ def get_subset_type(subset):
         return 'spectral'
     else:
         return None
+
+
+class MultiMaskSubsetState(SubsetState):
+    """
+    A subset state that can include a different mask for different datasets.
+    Adopted from https://github.com/glue-viz/glue/pull/2415
+
+    Parameters
+    ----------
+    masks : dict
+        A dictionary mapping data UUIDs to boolean arrays with the same
+        dimensions as the data arrays.
+    """
+
+    def __init__(self, masks=None):
+        super(MultiMaskSubsetState, self).__init__()
+        self._masks = masks
+
+    def to_mask(self, data, view=None):
+        if data.uuid in self._masks:
+            mask = self._masks[data.uuid]
+            if view is not None:
+                mask = mask[view]
+            return mask
+        else:
+            raise IncompatibleAttribute()
+
+    def copy(self):
+        return MultiMaskSubsetState(masks=self._masks)
+
+    def __gluestate__(self, context):
+        serialized = {key: context.do(value) for key, value in self._masks.items()}
+        return {'masks': serialized}
+
+    def total_masked_first_data(self):
+        first_data = next(iter(self._masks))
+        total_masked = [len(x) for x in np.where(self._masks[first_data])]
+        return total_masked[0]
```

Comment:
I think this is the same

```suggestion
        return len(np.where(self._masks[first_data])[0])
```